Reserve characters in names

mkitti opened the pull request:
This PR reserves characters in registered names.

The intention is to reserve characters to allow for namespace or a specific URI schemes in the future.

The reservations include all URI reserved characters as in RFC 3986.

They additionally include ".", "-", and "~". Underscores are allowed since `sharding_index` is already used.

xref: https://github.com/zarr-developers/zarr-specs/pull/330#issuecomment-2813003630

Capitalization should be considered in a future pull request. I recommend reserving all capital letters.

normanrz commented:
#2 introduces some codecs with `.` in their name. This is currently in use in zarr-python. Either we say `.` is a namespace delimiter and we assign the `numcodecs` prefix or we remove `.` from the reserved list, here.

Most of these characters are already prohibited by the `^[a-z][a-z0-9-_.]+$` regex in the spec: https://zarr-specs.readthedocs.io/en/latest/v3/core/index.html#id30

mkitti commented:
At the moment, I would lean towards the most conservative option possible.

For names with no prefix, I would tend to reserve `.` here so that we can use that to indicate prefixes.

Would those numcodecs extensions be registered here as well or do we delegate that namespace to be handled within the numcodecs repository?

I'm also a bit worried now about disambiguation with numcodecs.js versus the Python package.

mkitti commented:
I added a section on namespaces.

> ## Namespaces
> Namespaces MAY precede an extension name or other namespace prefixes. The character `.` MUST be used to delimit namespaces from each other and from the extension names. The primary use of namespaces is to disambiguate extensions with similar names which may differ in implementation or metadata. Extensions SHOULD register without a namespace if there is no amibiguity or difference in implementation from another extension of a similar name or use an existing namespace if applicable.

I also removed `-` as a reserved character since #5 is trying to use them. However, I am not sure why `-` is being used. I think we should discourage the use of `-` so that identifiers in common programming languages can match extension names exactly.
